src/tiledb/cloud/geospatial/ingestion.py

```diff
@@ -0,0 +1,1287 @@
+                    compressor_args = dict(compressor)
+                    compressor_name = compressor_args.pop("_name")
+                    if compressor_name:
+                        compressor_args = {
+                            k: None if not v else v for k, v in compressor_args.items()
+                        }
+                        compression_filter = vars(filter).get(compressor_name)(
+                            **compressor_args
+                        )
+                        compression_level = compression_filter.level
+
+                    else:
+                        raise ValueError
+
+                    with rasterio.open(
+                        dataset_uri,
+                        mode="w",
+                        COMPRESSION=compressor_name[
+                            :-6
+                        ].upper(),  # remove trailing `Filter`
+                        COMPRESSION_LEVEL=compression_level,
+                        **dst_profile,
+                    ):
+                        pass
+                else:
+                    with rasterio.open(
+                        dataset_uri,
+                        mode="w",
+                        **dst_profile,
+                    ):
+                        pass
+
+                logger.debug("Raster array created %r", dataset_uri)
+            else:
+                # srcs and dst have same number of bands
+                if resampling:
+                    resampling = rasterio.enums.Resampling[resampling.lower()]
+                with rasterio.open(dataset_uri, mode="r+") as dst:
+                    for chunk in chunks:
+                        input_datasets = [
+                            rasterio.open(f, opener=vfs.open) for f in chunk.files
+                        ]
+                        try:
+                            row_off = chunk.ranges[0][0]
+                            col_off = chunk.ranges[1][0]
+                            chunk_window = rasterio.windows.Window(
+                                col_off,
+                                row_off,
+                                chunk.ranges[1][1] - col_off,
+                                chunk.ranges[0][1] - row_off,
+                            )
+                            chunk_bounds = rasterio.windows.bounds(
+                                chunk_window, dst.transform
+                            )
+
+                            for b in range(dst.count):
+                                chunk_arr, _ = rasterio.merge.merge(
+                                    input_datasets,
+                                    bounds=chunk_bounds,
+                                    nodata=nodata,
+                                    indexes=[b + 1],
+                                )
+                                dst.write(chunk_arr, window=chunk_window)
+                        finally:
+                            map(lambda s: s.close(), input_datasets)
```

Comment:
Since this isn't really a functional transformation, I think that

```python
for s in input_datasets:
  s.close()
```

is probably a better representation for what's going on here.

Reply:
👍